firewalld_ipset module

rwaweber opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Adds a `firewalld_ipset` module, for the primary interest of being able to create simpler playbooks that interact with firewalld. This was inspired by having to maintain long, sprawling lists of firewall rules that ended up staying in a `with_items` and expanding my playbook execution times drastically.

With this, it becomes much faster (and hopefully simpler) to be able to group/manage those same lists into more organized groups of systems/network segments, while batching the individual allow/deny into one task.

Related: https://github.com/ansible/ansible/issues/37319
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`firewalld_ipset`
`firewalld`
##### ADDITIONAL INFORMATION
There were some neat abstractions for firewalld added in https://github.com/ansible/ansible/pull/37603. Unfortunately, I wasn't totally sure how to incorporate them here. I'd like to be able to use them here if it seems possible/helpful.

Also worth mentioning -- copied some of the docs notes from firewalld regarding the python client  and emulated the import statement to mostly bypass the lack of the client being installed for the doc tests.

Would also like to add some unit/integration tests as well, though unsure of a sane way considering the isolated use on linux and dependencies on iptables, firewalld, dbus, and likely systemd. A container probably isn't the best place to do something like that, right?

rwaweber committed:
First attempt at a documented firewalld_ipset module

rwaweber committed:
fix boilerplate and pep8

maxamillion commented on lib/ansible/modules/system/firewalld_ipset.py:
```diff
@@ -0,0 +1,267 @@
+  addresses:
+    description: The list of addresses to be added or removed from the ipset. This can be ipv4, ipv6, and mac addresses.
+    required: False
+    type: list
```
Is there any desire to support offline firewall operations like the `firewalld` module does?

maxamillion commented on lib/ansible/modules/system/firewalld_ipset.py:
```diff
@@ -0,0 +1,267 @@
+        argument_spec=module_args,
+        supports_check_mode=True
+    )
+
```
I would suggest using the firewalld `module_util` to help with the version sanity checking here.
```python
from ansible.module_utils.firewalld import FirewallTransaction
FirewallTransaction.sanity_check(module)
```

maxamillion commented on lib/ansible/modules/system/firewalld_ipset.py:
```diff
@@ -0,0 +1,267 @@
+        "firewalld_ipset_addresses": module.params['addresses']
+    }
+
+    if module.check_mode:
```
Check mode should check system state and return `changed=True` in the event something would have changed if the user were to have executed a task using this module not in check mode.

maxamillion commented on lib/ansible/modules/system/firewalld_ipset.py:
```diff
@@ -0,0 +1,267 @@
+    module.exit_json(**result)
+
+
+def main():
```
This seems redundant since you only call `run()` here and the in `if __name__ == '__main__`:` you only call `main()`. I'd suggest renaming `run` to `main`.

maxamillion reviewed (changes requested):
Thank you for the pull request and contribution of a new module!

I've added some review comments through out the code and would like to request the addition of some integration test cases similar to those found [here](https://github.com/ansible/ansible/blob/devel/test/integration/targets/firewalld/). Do note that since this is a new module, it should be defined as a new target.

rwaweber committed:
Incorporating suggestions:

- add version check from module_utils
- remove redudandant function call
- incorporate check_mode logic

maxamillion commented:
This looks good, thank you!

Would you mind adding some [integration tests](https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html) similar to those found for [firewalld](https://github.com/ansible/ansible/blob/devel/test/integration/targets/firewalld/tasks/service_test_cases.yml).

needs_info

rwaweber commented:
Definitely! Sorry I should’ve changed this to WIP.

On Thu, Dec 20, 2018 at 2:35 PM Adam Miller <notifications@github.com>
wrote:

> This looks good, thank you!
>
> Would you mind adding some integration tests
> <https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html>
> similar to those found for firewalld
> <https://github.com/ansible/ansible/blob/devel/test/integration/targets/firewalld/tasks/service_test_cases.yml>
> .
>
> needs_info
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/49850#issuecomment-449111583>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFkIGfcP3qDWO7BE0Cp5ubUIQW8vkA9Wks5u6-ZtgaJpZM4ZRCg0>
> .
>
-- 
Sent from my mobile device

rwaweber committed:
firewalld_ipset: add integration tests

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/105288/69/tests

rwaweber commented:
Thanks @mattclay! Looks like it might be failing on the latest versions of the enabled distributions, unless I specified something wrong in the testing targets.

Seems like  the `FirewallClient` init is returning a NoneType.

rwaweber commented:
After a little sleuthing of turning firewalld on and off on one of my own machines it definitely seems to be related to firewalld not running on the environment where the tests are being run.

I think I'm going to try to get some offline functionality baked into this. Unfortunately, it doesn't look like there is an out of the box way to integrate with the `firewall.core.fw_test.Firewall_test` object wrt ipsets.

I think in the same vein I'm going to try to reuse some more of the functionality that comes built into `ansible.module_utils.firewalld` to cover some of the easy edge cases rather than reimplementing the same checks.

maxamillion commented:
@rwaweber feel free to add functionality or otherwise make changes to `ansible.module_utils.firewalld` as well, that `module_util` exists to enable additional firewalld related modules such as this one so if it's not fulfilling that requirement then let's make it so. :)

rwaweber commented:
Absolutely! Given that there doesn't really seem to be offline capabilities within the `firewall.core.fw_test.Firewall_test` object, do you think it would be a _terrible_ idea to do xml mangling for the ipset files instead?

I may dig into whether or not there's some state to be managed inside of firewalld separately other than the files.

maxamillion commented:
@rwaweber I think I'd prefer to not do xml mangling but that's mostly because I greatly dislike doing such things, if you come up with a reasonable way to manage transactional state change to the xml files then I'm open to it. I'd be curious if there's internal firewalld libraries for such things.

elcomtik commented:
@rwaweber do you plan to continue working on this? I would really appreciate this feature? Do you want some help with finishing it?

rwaweber commented:
Hey @elcomtik! Sorry, this has totally fallen off my radar.

I think if I remeber correctly, one of the last lingering issues here was that there wasn't really a simple way to do offline state management for ipsets, mostly because it doesn't seem that the functionality was exposed within the destination firewalld library(totally possible that this has changed in the interim). The offline management features were needed to be able to put together some integration tests.

Thinking about it, if I were to come at this again, it might make more sense to approach this from the netfilter-ipset angle, rather than strictly from within a firewalld construct. That way, ipset lists can be made accessible to all downstream users of netfilter primitives, including firewalld, iptables, ufw, etc.

Unfortunately, I'm not familiar with the state of python bindings for that sort of tooling today. Though, I think the general approach used here could be used pretty easily there as well. In that, you're effectively just managing a named list of metadata and percolating out if/when that list changes.

I'm going to go ahead and close this issue as I don't plan to continue to work on this.

If you are looking for something that looks _kind_ of like this, I have an example playbook with a vendored version of this module that you're more than welcome to take a look at and adapt for your own use here in an archived repo of mine: https://github.com/rwaweber/firewalld-playbook.

elcomtik commented:
Thank @rwaweber for your suggestions, I will look at this.
